### PR TITLE
Feature raw websocket

### PR DESCRIPTION
**Pull request by michax**
This PR Implements plain/raw websocket transport with ability to fallback to old signalr support in case of any failures.

Supports both JSON format and new Protobuf one.

Check wiki for more details: https://wiki/display/STGO/OpenApi+Client%3A+Raw+Websocket

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,63 +1,66 @@
...
-    "mockdate": "2.0.2",
-    "nodemon": "1.18.7",
-    "protobufjs": "6.8.8",
-    "regenerator-runtime": "0.13.1",
-    "rimraf": "2.6.2",
-    "rollup": "0.67.4",
-    "rollup-plugin-babel": "4.0.3"
-  }
-}
+{
```
Comment (michax) [inline package.json]: Hmm not sure what happen here. Different new lines probably. Will check. Otherwise file is almost the same (only diff is one new library: long).

File: libs/jquery.signalR-2.0.3.js [hunk 2]
```diff
@@ -40,6 +40,13 @@
         throw new Error(resources.nojQuery);
     }
 
+    window.$signalrStats = {
```
Comment (michax) [inline libs/jquery.signalR-2.0.3.js]: Those changes will be dropped as soon as I finish profiling size and performance difference between signalR and new plain websocket.

**Comment by michax** [comment]
Currently fixing failed unit tests and adding new ones.

File: package.json (hunk 1 above)
Comment (lukeapage) [reply package.json]: I don't see a new library? its probably line endings

File: src/utils/uint64.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+const D = 1000000000;
+
+/**
+ * Converting array buffer (Uint8Array.buffer) storing uint64 integer to a string.
+ * Useful for message id parsing which is in uint64 format.
+ * @param buffer
+ * @returns {string} - The string representation of unsigned 64-bit integer.
+ */
+export function uint64ToStringLE(buffer) {
+    const h = buffer[4] + buffer[5] * 0x100 + buffer[6] * 0x10000 + buffer[7] * 0x1000000;
```
Comment (lukeapage) [inline src/utils/uint64.js]: this lot seems excessive.. do we need the message id? do we need it converted exactly like this?

Comment (michax) [reply src/utils/uint64.js]: Yes, we need messageId. It's mimicking signalR behavior when on reconnect last message id was appended to re-connection url query. 

OpenApi also explicitly specifies this as an requirement in new plain websocket lifecycle.

In short it allows to continue subscription from specific point. 
More details: https://wiki/display/DP/Plain+WebSocket+Streaming#PlainWebSocketStreaming-Reconnecting

As for format itself. Uint64 is not exactly JS friendly. BigInt is coming but for now we need to manually format it to decimal string representation.

However, there might be one improvement here to be done. Instead of formatting string decimal messageId on each update, we might consider keeping it as Uint8Array, until it's needed in string format upon reconnection. For that i need to review MessageId usage in GO. Initial checks shows that is heavily used in mocks and explicitly in notification code.
---
It might be also worth discussing with OpenApi messageId format and make it just plain string with size prefix. It would for sure make any external integration easier (outside openapi-clientlib).
---
As for alternative conversion approach. Initial version used Long library. Which was quite heavy. 
This util seems to be smallest uint64 to string conversion that i could came up based on stack overflow sample: https://stackoverflow.com/questions/45455979/byte-array-to-uint64-as-a-string
---
TLDR: I would keep it as it is and yes it's required. For Next phase i would discuss with openapi if we could use string or different more JS friendly format here.

File: src/openapi/streaming/connection/transport/signalr-transport.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+SignalrTransport.NAME = NAME;
+
+SignalrTransport.isSupported = function() {
+    return true;
+};
+
+SignalrTransport.prototype.isSupported = SignalrTransport.isSupported;
+
+SignalrTransport.prototype.stateChanged = function(callback) {
```
[the remark this answers is not in the document]
Comment (michax) [reply src/openapi/streaming/connection/transport/signalr-transport.js]: Yes, makes sense (similar thing was already done for connection). Will change.

File: src/openapi/streaming/connection/connection.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+const LOG_AREA = 'PlainWebSocketsTransport';
+const DEFAULT_TRANSPORTS = [WebsocketTransport, SignalrTransport];
+
+const TRANSPORT_NAME_MAP = {
+    // New, plain RAW WebSocket transport.
+    [WebsocketTransport.NAME]: WebsocketTransport,
+
+    // Backward compatible mapping to SignalR.
+    'webSockets': SignalrTransport,
```
[the remark this answers is not in the document]
Comment (michax) [reply src/openapi/streaming/connection/connection.js]: Nope, it's still using internal signalR transport switching. 

Why I decided to keep it?

- New Plain WebSocket might fail in authorization phase. In such case it doesn't mean WebSocket is not supported. So i think it matters to try and use webSocket from signalR with ability then to fallback to longPolling (internally managed by signalR).
- New Plain WebSocket might fail during handshake phase. Similar explanation to above case.
- Simplicity.

File: src/openapi/streaming/connection/connection.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+};
+
+const NOOP = () => {};
+
+const STATE_CREATED = 'connection-state-created';
+const STATE_STARTED = 'connection-state-started';
+const STATE_STOPPED = 'connection-state-stopped';
+
+function onTransportFail(error) {
+    log.error(LOG_AREA, 'Transport failed', { error });
```
[the remark this answers is not in the document]
Comment (michax) [reply src/openapi/streaming/connection/connection.js]: For initial release I would recommend using aggressive fallback to signalR and cover it with additional logs about fallback reasons. For example:
1. Failed due to handshake connection error response.
2. Failed due to handshake connection error.
3. Failed due to authorization failure.
4. Failed due to close with error code. 
5. Failed due to lack of supported types (Uint8Array, WebSocket, DataView).

We could analyse results and act accordingly for Phase 2.

However if you believe that current logs already paint a relatively clear picture I would suggest following. 
- If failures are due to handshake or authorization issues, it could mean connection failures. In such scenarios try to reconnect within same transport.
- If failure is due to missing support ie. lack of WebSocket class, than fallback to signalR. 

Anyway, I wouldn't recommend dropping signalr in first plain/raw websocket release. However I would also love to drop it in the end.

File: src/openapi/streaming/connection/transport/websocket-transport.js [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        handleFailure.call(this, { message: 'Failed to setup webSocket connection' });
+    }
+}
+
+function parseMessage(rawData) {
+    let index = 0;
+    const messages = [];
+
+    while (index < rawData.byteLength) {
+        const message = new DataView(rawData);
```
[the remark this answers is not in the document]
Comment (michax) [reply src/openapi/streaming/connection/transport/websocket-transport.js]: Was planning to add it but first wanted to discuss what should happen when parsing error occurs. 

Should We:
1. Immediately fail transport and move to next supported one.
2. Set error threshold and when reached fallback to another transport.
3. Log error and carry on. 

I would vote for 2 option, with threshold set to 50. We could also require threshold to be hit within specific timeframe limit.

File: src/openapi/streaming/connection/connection.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukeapage) [reply src/openapi/streaming/connection/connection.js]: but when you start signal-r you specify both transports even though here its only one? does that means it effectively tries signal-r twice in the exact same way?
I guess it doesn't matter much if we can in the long run get rid of signal-r

File: src/openapi/streaming/connection/connection.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lukeapage) [reply src/openapi/streaming/connection/connection.js]: What about, fully connected, getting data, then gets disconnected?

File: src/openapi/streaming/connection/transport/websocket-transport.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lukeapage) [reply src/openapi/streaming/connection/transport/websocket-transport.js]: Lets go with (1) to start with - Its probably good to be conservative. Then we can see what errors we get on live.

File: src/utils/uint64.js (hunk 3 above)
Comment (lukeapage) [reply src/utils/uint64.js]: I'm not convinced the notifications messageId is the same thing - are you sure?

I like the idea of keeping it in a more raw format and only using it when required.

File: src/openapi/streaming/connection/connection.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (michax) [reply src/openapi/streaming/connection/connection.js]: Disconnect (onclose) should start reconnect loop. However it's highly probable that next connect attempt will fail, which in effect will call onerror and then transport will fail and cause fallback.

I like the idea of trying to reconnect more aggressively if transport worked at least once! Will implement.
